### ISSUE TITLE
API tooling reports wrong errors on Java 21

### ISSUE THREAD

**Issue by iloveeclipse** [eclipse-pde/eclipse.pde#771]
See errors reported by API tooling on https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1439 and https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1436.

```
15:18:26  3 API ERRORS
15:18:26  [API ERROR] File MANIFEST.MF at line 5: The major version should be incremented in version 3.36.0, since API breakage occurred since version 3.35.0 - The super interfaces set has been reduced (java.lang.constant.Constable) for type org.eclipse.jdt.core.util.ExternalAnnotationUtil.MergeStrategy
15:18:26  - The super interfaces set has been reduced (java.lang.constant.Constable) for type org.eclipse.jdt.core.dom.rewrite.ImportRewrite.TypeLocation (location: /home/jenkins/agent/workspace/eclipse.jdt.core-Github_PR-1436/org.eclipse.jdt.core/META-INF/MANIFEST.MF)
15:18:26  [API ERROR] File ImportRewrite.java at line 116: The super interfaces set has been reduced (java.lang.constant.Constable) for type org.eclipse.jdt.core.dom.rewrite.ImportRewrite.TypeLocation (location: /home/jenkins/agent/workspace/eclipse.jdt.core-Github_PR-1436/org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/rewrite/ImportRewrite.java)
15:18:26  [API ERROR] File ExternalAnnotationUtil.java at line 68: The super interfaces set has been reduced (java.lang.constant.Constable) for type org.eclipse.jdt.core.util.ExternalAnnotationUtil.MergeStrategy (location: /home/jenkins/agent/workspace/eclipse.jdt.core-Github_PR-1436/org.eclipse.jdt.core/model/org/eclipse/jdt/core/util/ExternalAnnotationUtil.java)
```

They both run on Java 21 on Jenkins and both show same API errors, all reported on enums defined in JDT.

As I've wrote on https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1436, I suspect a problem with classpath resolving in API tooling.

It looks like if the API baseline would see java.lang.Enum from Java 8 or 11 in the **current** API (why???) and compare it with java.lang.Enum from Java 17 **in the API baseline** where it **additionally** implements Constable (since Java 12).

So it would be as if PDE is not able to recognize Java 21 as valid JDK and fails back to Java 11 (or 8). 
This test from https://github.com/eclipse-pde/eclipse.pde/issues/764 complains that `Classpath entry for JavaSE-17 is missing` which is the reason why I'm assuming we have a bug in PDE.

JDT tests itself run fine (in most cases) on Java 21, we only have one known test fail but that seem to be unrelated.

**Comment by iloveeclipse** [comment]
OK, still same errors after #772 : https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-1439/3/console

```
17:11:36  [INFO] --- tycho-apitools:4.0.2:verify (verify) @ org.eclipse.jdt.core ---
17:11:36  [INFO] Resolve API baseline for org.eclipse.jdt:org.eclipse.jdt.core:eclipse-plugin:3.36.0-SNAPSHOT
17:11:36  [INFO] Adding repository https://download.eclipse.org/eclipse/updates/4.29/R-4.29-202309031000
17:11:38  [INFO] Resolve API tools runtime from MavenRepositoryLocation [location=https://download.eclipse.org/eclipse/updates/4.30-I-builds/]...
17:12:25  3 API ERRORS
17:12:25  [API ERROR] File MANIFEST.MF at line 5: The major version should be incremented in version 3.36.0, since API breakage occurred since version 3.35.0 - The super interfaces set has been reduced (java.lang.constant.Constable) for type org.eclipse.jdt.core.util.ExternalAnnotationUtil.MergeStrategy
17:12:25  - The super interfaces set has been reduced (java.lang.constant.Constable) for type org.eclipse.jdt.core.dom.rewrite.ImportRewrite.TypeLocation (location: /home/jenkins/agent/workspace/eclipse.jdt.core-Github_PR-1439/org.eclipse.jdt.core/META-INF/MANIFEST.MF)
17:12:25  [API ERROR] File ImportRewrite.java at line 116: The super interfaces set has been reduced (java.lang.constant.Constable) for type org.eclipse.jdt.core.dom.rewrite.ImportRewrite.TypeLocation (location: /home/jenkins/agent/workspace/eclipse.jdt.core-Github_PR-1439/org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/rewrite/ImportRewrite.java)
17:12:25  [API ERROR] File ExternalAnnotationUtil.java at line 68: The super interfaces set has been reduced (java.lang.constant.Constable) for type org.eclipse.jdt.core.util.ExternalAnnotationUtil.MergeStrategy (location: /home/jenkins/agent/workspace/eclipse.jdt.core-Github_PR-1439/org.eclipse.jdt.core/model/org/eclipse/jdt/core/util/ExternalAnnotationUtil.java)
17:12:25  24 API warnings
```

@laeubi : is it possible to know or log which exact API tooling version is used?

```
Resolve API tools runtime from MavenRepositoryLocation [location=https://download.eclipse.org/eclipse/updates/4.30-I-builds
```

doesn't say which of the builds is taken. I **hope** it is https://download.eclipse.org/eclipse/updates/4.30-I-builds/I20230927-0600/ but it would be nice to **know**.

Also would be nice to save error log from API tooling somewhere on Jenkins, just in case there are some exceptions reported.

**Comment by iloveeclipse** [comment]
Even worse, API check *sometimes* passes and *sometimes* fails... Just retriggered it and it failed on one PR  and passed on other one.

**Comment by vik-chand** [comment]
> Even worse, API check _sometimes_ passes and _sometimes_ fails... Just retriggered it and it failed on one PR and passed on other one.

Does this recreate in a workspace?

**Comment by iloveeclipse** [comment]
No. I've tried but it never failed. 

Meanwhile I assume we could have either race conditions with PDE / JDT init that happens only on very resource limited or slow systems (like JDT Jenkins that suffered last days from extreme slow Jenkins performance), or we have problem with tycho not using latest SDK IBuild to run tooling from, because of same extra unreliable http server access from Jenkins (may be related to https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/3672).

Also we might have the problem with "JDK autodetection" job that was added recently and caused various test fails in JDT until it was disabled. @laeubi : can we disable this job in tycho ? See https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1150

**Comment by laeubi** [comment]
> is it possible to know or log which exact API tooling version is used?

API tools do not really have a dedicated version (beside the bundle version where there could be multiple ones), but you can run the project locally for example and enable debug output to print out more details.

> Also would be nice to save error log from API tooling somewhere on Jenkins

What do you mean by error log? The Eclipse Log? I think one can register a LogListener and report any errors, the "ApiToolsCheck" is effectively implemented here where it already reports error markers on the project:

https://github.com/eclipse-pde/eclipse.pde/blob/master/apitools/org.eclipse.pde.api.tools/src/org/eclipse/pde/api/tools/internal/ApiAnalysisApplication.java

> Meanwhile I assume we could have either race conditions with PDE / JDT init that happens only on very resource limited or slow systems

Maybe, one problem is that there might be some *intermediate* state where there are failures that would later be resolved. I tried to fix this with waiting for all autobuilds but this seem not to be sufficient:

https://github.com/eclipse-pde/eclipse.pde/blob/d8a3ad41fdcfeda6ab6c0db1572c82663058c43e/apitools/org.eclipse.pde.api.tools/src/org/eclipse/pde/api/tools/internal/ApiAnalysisApplication.java#L132

so if someone has a better idea on how to make sure "everything is done" it would be great to propose a patch as the new automatic manifest feature seem to suffer from the same problem.

Another (even better) approach would be to add an option to not compile the project at all, this would also speed up things a lot as Tycho already has the jar as a project maybe @vik-chand can help here how to modify the `ApiAnalysisApplication` that it accepts a jar instead of a project?

**Comment by laeubi** [comment]
> can we disable this job in tycho

The ApiChecks? I don't think that my choice here, but the [-Papi-check](https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/blob/c56b459b8c4a1acc9eae8e875dfc3a67565fd95a/eclipse-platform-parent/pom.xml#L779C12-L779C21) controls if API checks are to be performed in the buildscripts.

**Comment by iloveeclipse** [comment]
> > can we disable this job in tycho
> 
> The ApiChecks? 

No! The "auto detect JDK" job, see above

**Comment by laeubi** [comment]
> No! The "auto detect JDK" job, see above

Sure actually the `ApiAnalysisApplication` is only used by platform/tycho so I don't see a reason to not change it, if you know how to do that just open a PR and it should be picket up after the next I-Build.

**Comment by Bananeweizen** [comment]
I understood that you only saw this on Jenkins until now. I have that error in my egit workspace, would that be helpful for debugging?
![grafik](https://github.com/eclipse-pde/eclipse.pde/assets/406876/19341e3f-ac74-446d-b2ee-44fbca0923ab)

**Comment by iloveeclipse** [comment]
I assume it is a duplicate of https://github.com/eclipse-pde/eclipse.pde/issues/1073 and was fixed via https://github.com/eclipse-pde/eclipse.pde/pull/1093
